### ISSUE TITLE
 Build and Run QNNPACK on X86

### ISSUE THREAD

**Issue by Darshvino** [pytorch/pytorch#83509]
### 🚀 The feature, motivation and pitch

Hi Pytorch team,

I want to build and run QNNPACK on X86 CPU and there are not many references to do the same. It would be great if I can get the references to build and execute QNNPACK on the X86 platform.

Thank you!

**Comment by dhruvbird** [comment]
cc @kimishpatel @SS-JIA @digantdesai for guidance on building QNNPACK on x86 CPU. I think it should be possible, and it is already done within the PyTorch build.

**Comment by digantdesai** [comment]
Yeah, it should be there for x86 builds. I.e. for an x86 build, you should be able to set quantized backed to `qnnpack`.

**Comment by malfet** [comment]
Just to clarify - you are asking for 32-bit build, aren't you? 
And aren't QNNPACK archived?

**Comment by kimishpatel** [comment]
Yes pytorch/qnnpack repo is archived and not maintained anymore.

**Comment by kimishpatel** [comment]
I think the issue needs more clarification. Do you want to build qnnpack from https://github.com/pytorch/qnnpack? Pytorch uses hard fork of qnnpack inside https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu/qnnpack. So if you want to build pytorch with qnnpack then what @digantdesai said holds true.

**Comment by Darshvino** [comment]
Hi @kimishpatel @digantdesai @malfet,

Thank you for reply.

I can see there is already a qnnpack backend available with pip install.

<img width="343" alt="image" src="https://user-images.githubusercontent.com/41161376/224206318-a3e87ada-790a-4c0f-8cd7-7ff5bccfada6.png">

My current doubt is how can we select qnnpack as a backend for quantized inference on X86 CPU?

Actually, I tweaked the QNNPACK GEMM kernel to implement the custom algorithm: https://github.com/pytorch/pytorch/blob/5f89d147a100064ae8a9743fa0e297bdedfa1b61/aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/2x4c8-sse2.c#L41 and I want to make sure that that kernel runs while running a CNN op.

Also, where can I get info in the code repo about selecting a particular GEMM kernel under QNNPACK for execution on X86 CPU. basically I want to know the strategy of execution of the kernels, so I can make sure that the custom kernel with the **pytorch_q8gemm_ukernel_2x4c8__sse2** can be executed.

Thanks

**Comment by Darshvino** [comment]
Hi @kimishpatel @digantdesai @malfet,

Look forward to your reply.

Thanks

**Comment by kimishpatel** [comment]
> Also, where can I get info in the code repo about selecting a particular GEMM kernel under QNNPACK for execution on X86 CPU. basically I want to know the strategy of execution of the kernels, so I can make sure that the custom kernel with the pytorch_q8gemm_ukernel_2x4c8__sse2 can be executed.

For this you can just throw or assert in the code to make sure code line is reached. As for the actual selection, it happens here https://github.com/pytorch/pytorch/blob/5f89d147a100064ae8a9743fa0e297bdedfa1b61/aten/src/ATen/native/quantized/cpu/qnnpack/src/init.c#L55. But I see that it does not reference kernel you are touching.
